xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java

```diff
@@ -492,9 +493,7 @@ public String toString()
      */
     private Map<DocumentReference, List<BaseObject>> xObjects = new TreeMap<DocumentReference, List<BaseObject>>();
 
-    // TODO: use a Map instead of a List to store attachment in XWikiDocument
-    private final List<XWikiAttachment> attachmentList =
-        new AbstractNotifyOnUpdateList<XWikiAttachment>(new ArrayList<XWikiAttachment>())
+    private final List<XWikiAttachment> attachmentList = new XWikiAttachmentList()
         {
             @Override
             public void onUpdate()
```

Comment:
I would have also moved those methods to XWikiAttachmentList (setMetaDataDirty is public so you can call it from another class). You could pass the XWikiDocument instance to the XWikiAttachmentList constructor for example.